google-cloud-storage/src/main/java/com/google/cloud/storage/RestartableInputStream.java

```diff
@@ -0,0 +1,48 @@
+/*
+ * Copyright 2015 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.storage;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+
+/**
+ *
+ */
+public class RestartableInputStream {
+
+    InputStream inputStream;
+
+    public RestartableInputStream(InputStream inputStream) {
+        this.inputStream = inputStream;
+        this.inputStream.mark(Integer.MAX_VALUE);
```

Comment:
Can you please share you basic idea/design here? Using `.mark()` seems promising but it is not as straightforward. For example, not each `InputStream` supports  `mark/reset` methods, and the default implementation does not support it: [java.io.InputStream.markSupported()](http://grepcode.com/file/repository.grepcode.com/java/root/jdk/openjdk/6-b14/java/io/InputStream.java#367).

Reply:
The idea here is using `mark()` to mark the start of the `InputStream` so that when retrying, calling `reset()` would make the stream go back to where we want to start. I was not quite confident with this design here either - so besides mark()/reset(), is there a better way to cache the bytes that have been read previously?